### PR TITLE
Collision forces and associated testcases

### LINKED ISSUE #1836: Implementation collision forces for moving rigid bodies in fluids
Collision forces are necessary to avoid contact between moving objects in fluids
they can be implemented as a plugin for the fluid toolbox

### LINKED ISSUE #1837: add moving rigid body testcases in fluids and including collision tests
minimal testcases should be built to verify the proper behavior of the motion of the rigid bodies as well as the case of collision

### LINKED ISSUE #1839: Cleanup and organize testcases on  bodies moving in fluids
There are too many testcases on moving body motion. We should keep the ones that allow to check properly (norms measures ...)  the toolbox features

### PR DESCRIPTION
**Pull request by prudhomm**
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/feelpp/feelpp/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes?
- [x] Have you successfully run the Feel++ testsuite with your changes locally?
- [ ] Have you written Doxygen comments in your contribution ?

-----

- closes #1836
- closes #1837
- closes #1839

### CODE REVIEW

**Comment by vincentchabannes** [comment]
@prudhomm it should go on feelpp/toolbox repo. There are already a lot of example with rigid_body/swimmer and I can't maintain all. We should select only few of them usefull for test

**Comment by prudhomm** [comment]
@vincentchabannes there will be some cleanup but we need to have true testcases including collision where we verify the results automatically. I don't believe we have any checks on that matter.
Tests that are not checking anything will be moved butwe need to ensure the testing of toolbox features

**Comment by prudhomm** [comment]
I have retitled the PR

**Comment by vincentchabannes** [comment]
one case is enough,  and naca testcase is not a collision test case.

**Comment by prudhomm** [comment]
@VLCeline is the python wrapper fluid.cpp and fluid-remesh.py up to date in this branch or do you have changes still sitting on your gaya account ?

**Comment by prudhomm** [comment]
> one case is enough, and naca testcase is not a collision test case.

it will probably go in feelpp/toolboxes since it is not even in the paper

**Comment by VLCeline** [comment]
> @VLCeline is the python wrapper fluid.cpp and fluid-remesh.py up to date in this branch or do you have changes still sitting on your gaya account ?

I have some changes on my gaya account.

**Comment by lberti** [comment]
@prudhomm @VLCeline checks on this branch have failed.
Also, the branch seems outdated with respect to develop.
Is anyone working on fixing this or should I do it?

**Comment by prudhomm** [comment]
I am not. Thank you for taking care of this

**Comment by prudhomm** [comment]
@vincentchabannes could you review this PR ? the extra tests have been removed by @lberti 
@VLCeline will add 2 testcases for the collision model

File: python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/contactforce.hpp [hunk 1]
```diff
@@ -0,0 +1,853 @@
...
+#include <feel/feelalg/glas.hpp>
+#include <feel/feelcore/json.hpp>
+#include <fmt/core.h>
+#include <Eigen/Geometry> 
+#include <feel/feelfilters/exporter.hpp>
+#include <feel/feells/distancetorange.hpp>
+#include <feel/feelcore/ptreetools.hpp>
+#include <feel/feeldiscr/minmax.hpp>
+#include <feel/feelvf/vf.hpp>
+#include <mpi.h>
```
Comment (vincentchabannes) [inline python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/contactforce.hpp]: remove mpi include and maybe check which are useless

File: python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/contactforce.hpp [hunk 2]
```diff
@@ -0,0 +1,853 @@
+#include <feel/feelmodels/modelcore/modelnumerical.hpp>
+#include <feel/feelmodels/fluid/fluidmechanics.hpp>
+#include <feel/feelalg/glas.hpp>
+#include <feel/feelcore/json.hpp>
+#include <fmt/core.h>
+#include <Eigen/Geometry> 
+#include <feel/feelfilters/exporter.hpp>
+#include <feel/feells/distancetorange.hpp>
+#include <feel/feelcore/ptreetools.hpp>
```
Comment (vincentchabannes) [inline python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/contactforce.hpp]: ptree is deprecated  (now we nl::json)

File: python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/fluid.cpp [hunk 3]
```diff
@@ -26,6 +26,9 @@
 #include <feel/feelpython/pybind11/functional.h>
 #include <feel/feelmodels/modelcore/modelnumerical.hpp>
 #include <feel/feelmodels/fluid/fluidmechanics.hpp>
+#include <feel/feelmodels/modelcore/remeshinterpolation.hpp>
+#include <feel/feelcore/pybind11_json.hpp>
+#include "contactforce.hpp"
```
Comment (vincentchabannes) [inline python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/fluid.cpp]: use < >

File: toolboxes/fluid/cases/rigid_body_paper_tests/falling_ellipse/falling_ellipse.cfg [hunk 4]
```diff
@@ -1,37 +1,39 @@
...
-time-step=0.1#23
-time-final=20.1
-#restart=true
+order=2
+time-step=0.001# the time step should be verified
+time-final=2
 restart.at-last-save=true
-time-initial=-0.1#0#-0.1
-#save.freq=2
+time-initial=-0.001
```
Comment (vincentchabannes) [inline toolboxes/fluid/cases/rigid_body_paper_tests/falling_ellipse/falling_ellipse.cfg]: time initial is negatif? should be probably 0 (and not depends on the time step)

File: toolboxes/fluid/cases/rigid_body_paper_tests/naca0012_profile/naca0012_profile.cfg [hunk 5]
```diff
@@ -0,0 +1,67 @@
+
+directory=toolboxes/fluid/new_testcases/naca0012_profile
```
Comment (vincentchabannes) [inline toolboxes/fluid/cases/rigid_body_paper_tests/naca0012_profile/naca0012_profile.cfg]: why add new_testcases?  should be : toolboxes/fluid/moving_body/naca0012_profile

File: toolboxes/fluid/cases/rigid_body_paper_tests/falling_ellipse/falling_ellipse.cfg [hunk 6]
```diff
@@ -1,37 +1,39 @@
-directory=toolboxes/fluid/moving_body/three_sphere_2D/
+
+directory=toolboxes/fluid/new_testcases/falling_ellipse
```
Comment (vincentchabannes) [inline toolboxes/fluid/cases/rigid_body_paper_tests/falling_ellipse/falling_ellipse.cfg]: why add new_testcases? should be : toolboxes/fluid/moving_body/falling_ellipse

**Review by vincentchabannes — CHANGES_REQUESTED** [review]
You should also remove .DStore and mv cases in rigid_body_paper_tests in moving_body directory (already exist)

File: python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/fluid.cpp (hunk 3 above)
Comment (VLCeline) [reply python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/fluid.cpp]: Where can I put the file `contactforce.hpp` (contains the collision algorithm) ? in which directory ?

File: toolboxes/fluid/cases/rigid_body_paper_tests/falling_ellipse/falling_ellipse.cfg (hunk 4 above)
Comment (lberti) [reply toolboxes/fluid/cases/rigid_body_paper_tests/falling_ellipse/falling_ellipse.cfg]: fixed

File: toolboxes/fluid/cases/rigid_body_paper_tests/naca0012_profile/naca0012_profile.cfg (hunk 5 above)
Comment (lberti) [reply toolboxes/fluid/cases/rigid_body_paper_tests/naca0012_profile/naca0012_profile.cfg]: fixed

File: toolboxes/fluid/cases/rigid_body_paper_tests/falling_ellipse/falling_ellipse.cfg (hunk 6 above)
Comment (lberti) [reply toolboxes/fluid/cases/rigid_body_paper_tests/falling_ellipse/falling_ellipse.cfg]: fixed

File: python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/contactforce.hpp (hunk 2 above)
Comment (lberti) [reply python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/contactforce.hpp]: fixed

File: python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/fluid.cpp (hunk 3 above)
Comment (VLCeline) [reply python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/fluid.cpp]: I used `include_directories` in the CMakeList to include the` contactforce.hpp` file. I couldn't use `target_include_directories` with the macro `feelpp_add_pymodule`.

**Comment by lberti** [comment]
@vincentchabannes we have made the changes there were demanded. The buildkite check has passed as well.

**Review by vincentchabannes — APPROVED** [review]

**Comment by vincentchabannes** [comment]
WARNING, this testcases can NOT be used with feelpp_toolbox_fluid executable or standard python script. Currently, we can use collision model only in python by using this kind of script `python/pyfeelpp-toolboxes/tests/test_collision.py` @lberti @VLCeline @prudhomm is it correct?

**Pull request merged**